nixos/modules/services/web-apps/mastodon.nix

```diff
@@ -0,0 +1,339 @@
+        '';
+        type = lib.types.str;
+      };
+      smtpPasswordFile = lib.mkOption {
+        description = ''
+          Path to file containing the SMTP password
+        '';
+        type = lib.types.str;
+      };
+
+      redisHost = lib.mkOption {
+        description = "Redis host";
+        type = lib.types.str;
+        default = "127.0.0.1";
+      };
+      redisPort = lib.mkOption {
+        description = "Redis port";
+        type = lib.types.port;
+        default = 6379;
+      };
+      dbHost = lib.mkOption {
+        description = "Postgres database host";
+        type = lib.types.str;
+        default = "127.0.0.1";
+      };
+      dbPort = lib.mkOption {
+        description = "Postgres database port";
+        type = lib.types.port;
+        default = 5432;
+      };
+      dbName = lib.mkOption {
+        description = "Postgres database name";
+        type = lib.types.str;
+        default = "mastodon";
+      };
+      localDomain = lib.mkOption {
+        description = "The domain serving your Mastodon instance";
+        default = "social.example.org";
+        type = lib.types.str;
+      };
+      smtpServer = lib.mkOption {
+        description = "SMTP host used when sending Emails to users";
+        type = lib.types.str;
+      };
+      smtpPort = lib.mkOption {
+        description = "SMTP port used when sending Emails to users";
+        type = lib.types.port;
+        default = 587;
+      };
+      smtpFromAddress = lib.mkOption {
+        description = ''"From" address used when sending Emails to users'';
+        type = lib.types.str;
+      };
+      elasticsearchHost = lib.mkOption {
+        description = ''
+          Elasticsearch host
+          If it is not null, Elasticsearch full text search will be enabled
+        '';
+        type = lib.types.nullOr lib.types.str;
+        default = null;
+      };
+      elasticsearchPort = lib.mkOption {
+        description = "Elasticsearch port";
+        type = lib.types.port;
+        default = 9200;
+      };
+
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    systemd.services.mastodon-init-dirs = {
+      script = ''
+        umask 077
+        cat > /var/lib/mastodon/.secrets_env <<EOF
+        SECRET_KEY_BASE=$(cat ${cfg.secretKeyBaseFile})
+        OTP_SECRET=$(cat ${cfg.otpSecretFile})
+        VAPID_PRIVATE_KEY=$(cat ${cfg.vapidPrivateKeyFile})
+        DB_PASS=$(cat ${cfg.dbPassFile})
+        SMTP_PASSWORD=$(cat ${cfg.smtpPasswordFile})
+        EOF
+      '';
+      serviceConfig = {
+        Type = "oneshot";
+        User = cfg.user;
+        Group = cfg.group;
+        LogsDirectory = "mastodon";
+        StateDirectory = "mastodon";
+      };
+      after = [ "network.target" ];
+      wantedBy = [ "multi-user.target" ];
+    };
+
+    systemd.services.mastodon-init-db = {
+      script = ''
+        cd ${pkgs.mastodon}
+        rake db:migrate
```

Comment:
Note that if this does automatic database migration, NixOS rollbacks might not work

Reply:
I've been wondering myself how this should be handled best but haven't come to a good conclusion

Reply:
I think this is a question of state management, which is not really dealt with by Nix. The same problem exists for all stateful data on the system (it can not be rolled back unless the user has a backup). Maybe a note in the docs would be good to remind users of that.

Reply:
As soon as tootctl works, I will add a "cache clear" call. This should be sufficient for 99% of the releases. The mastodon releases page lists the following steps for a major release:
 - Install dependencies - done by Nix
 - Run database migrations - done by mastodon-init-db.service
 - Precompile the assets - done by Nix
 - Restart all Mastodon processes - done by the activate script
 - Clear cache - this is what we are still missing

Reply:
I thought the migrate task worked fine for an empty database on my deployment. I would say I will take a look but I'm pretty busy. If it doesn't work for an empty database, we will of course need that.

Reply:
Migrating an empty DB will be slow and error prone, if nothing else, since it has to construct it migration by migration and some of those migrations may not have been touched for years.

Reply:
I'm really uneasy about doing migrations automatically, even more when no automatic backups is done before.

I wish there could be an option to disable it.
(Btw, thanks for your work <3 !)

Reply:
Sure, I can add an option to opt-out.

Reply:
added an automaticMigrations option